typescript/sdk/src/token/adapters/EvmTokenAdapter.ts

```diff
@@ -326,18 +332,85 @@ export class EvmHypCollateralAdapter
     return this.collateralContract.allowedBridges(domain, bridge);
   }
 
-  async populateRebalanceTx(params: {
-    domain: Domain;
-    amount: Numberish;
-    bridge: Address;
-  }): Promise<PopulatedTransaction> {
-    const interchainGas = await this.quoteTransferRemoteGas(params.domain);
+  async getRebalanceQuotes(
+    bridge: Address,
+    domain: Domain,
+    recipient: Address,
+    amount: Numberish,
+    isWarp: boolean,
+  ): Promise<InterchainGasQuote[]> {
+    // TODO: In the future, all bridges should get quotes from the quoteTransferRemote function.
+    // Given that currently warp routes used as bridges do not, quotes need to be obtained differently.
+    // This can probably be removed in the future.
+    if (isWarp) {
+      const gasRouter = GasRouter__factory.connect(bridge, this.getProvider());
+      const gasPayment = await gasRouter.quoteGasPayment(domain);
+
+      return [
+        {
+          amount: BigInt(gasPayment.toString()),
+        },
+      ];
+    }
+
+    const bridgeContract = ValueTransferBridge__factory.connect(
+      bridge,
+      this.getProvider(),
+    );
+
+    const quotes = await bridgeContract.quoteTransferRemote(
+      domain,
+      addressToBytes32(recipient),
+      amount,
+    );
+
+    return quotes.map((quote) => ({
+      addressOrDenom:
+        quote.token === ethersConstants.AddressZero ? undefined : quote.token,
+      amount: BigInt(quote.amount.toString()),
+    }));
+  }
+
+  /**
+   * @param quotes - The quotes returned by getRebalanceQuotes
+   */
+  populateRebalanceTx(
+    domain: Domain,
+    amount: Numberish,
+    bridge: Address,
+    quotes: InterchainGasQuote[],
+  ): Promise<PopulatedTransaction> {
+    return this.basePopulateRebalanceTx(
+      domain,
+      amount,
+      bridge,
+      quotes,
+      () => quotes[0].amount,
+    );
+  }
+
+  protected async basePopulateRebalanceTx(
+    domain: Domain,
+    amount: Numberish,
+    bridge: Address,
+    quotes: InterchainGasQuote[],
+    getTrxValue: () => bigint,
```

Comment:
why defer this value to a function call?

Reply:
One way I found to avoid duplicating code for both Collateral and Native adapters

Reply:
Native inherits Collateral